packages/core/src/typings/openapi.ts

```diff
@@ -1,19 +1,26 @@
-export interface Oas3Definition {
+// common fields for OAS descriptions v3.x
+interface Oas3DefinitionBase<T extends Oas3Schema | Oas3_1Schema> {
   openapi: string;
   info?: Oas3Info;
   servers?: Oas3Server[];
-  paths?: Oas3Paths;
-  components?: Oas3Components;
+  paths?: Oas3Paths<T>;
+  components?: T extends Oas3_1Schema ? Oas3_1Components : Oas3ComponentsBase<T>;
```

Comment:
Related to https://github.com/Redocly/redocly-cli/pull/1795/files?diff=split&w=1#r1898410733 :

```suggestion
  components?: T extends Oas3_1Schema ? Oas3_1Components : Oas3Components;
```